File: BlueSTSDK/src/main/java/com/st/BlueSTSDK/Utils/advertise/AdvertiseParser.java
package com.st.BlueSTSDK.Utils.advertise;

import android.util.SparseArray;

public class AdvertiseParser {

    public final static byte FLAG_DATA_TYPE = 0x01;
    public final static byte INCOMPLETE_LIST_OF_128_UUID = 0x06;
    public final static byte DEVICE_NAME_TYPE = 0x09;
    public final static byte TX_POWER_TYPE = 0x0A;
    public final static byte VENDOR_DATA_TYPE = (byte) 0xff;

    public static SparseArray<byte[]> split(byte advertise[]){
        SparseArray<byte[]> splitAdvertise = new SparseArray<>();
        int ptr = 0;
        while (ptr < advertise.length - 2) {
            int length = advertise[ptr++] & 0xff;
            if (length == 0)
                break;

            final byte type = (advertise[ptr++]);
            //min between the length field and the remaining array length
            final int fieldLength = Math.min(length-1,advertise.length-ptr);

            byte data[] = new byte[fieldLength];
            System.arraycopy(advertise,ptr,data,0,fieldLength);
            splitAdvertise.put(type,data);

            ptr += fieldLength;
        }
        return splitAdvertise;
    }

}

File: st_blue_sdk/src/main/java/com/st/blue_sdk/bt/advertise/AdvertiseParser.kt
/*
 * Copyright (c) 2022(-0001) STMicroelectronics.
 * All rights reserved.
 * This software is licensed under terms that can be found in the LICENSE file in
 * the root directory of this software component.
 * If no LICENSE file comes with this software, it is provided AS-IS.
 */
package com.st.blue_sdk.bt.advertise

import android.util.SparseArray
import kotlin.math.min

object AdvertiseParser {

    fun split(advertise: ByteArray): SparseArray<ByteArray> {
        val splitAdvertise = SparseArray<ByteArray>()
        var ptr = 0
        while (ptr < advertise.size - 2) {

            val length: Int = advertise[ptr++].toInt() and 0xff

            if (length == 0)
                break

            val type = advertise[ptr++]

            //min between the length field and the remaining array length
            val fieldLength = min(length - 1, advertise.size - ptr)
            val data = ByteArray(fieldLength)
            System.arraycopy(advertise, ptr, data, 0, fieldLength)
            splitAdvertise.put(type.toInt(), data)
            ptr += fieldLength
        }
        return splitAdvertise
    }
}